packages/http/src/mocker/negotiator/NegotiatorHelpers.ts

```diff
@@ -53,236 +60,283 @@ function findResponseByStatusCode(
+      // OR
+      // a httpContent for a provided mediaType does not exist
+      logger.trace('No mediaType provided. Fallbacking to the default media type (application/json)');
+      return helpers.negotiateDefaultMediaType(
+        {
+          code,
+          dynamic,
+          exampleKey,
+        },
+        response,
+      );
+    });
   },
 
   negotiateOptionsForDefaultCode(
     httpOperation: IHttpOperation,
     desiredOptions: NegotiationOptions,
-  ): IHttpNegotiationResult {
+  ): Reader<Logger, Either<Error, IHttpNegotiationResult>> {
     const lowest2xxResponse = findLowest2xx(httpOperation.responses);
     if (lowest2xxResponse) {
       return helpers.negotiateOptionsBySpecificResponse(httpOperation, desiredOptions, lowest2xxResponse);
     }
-    throw new Error('No 2** response defined, cannot mock');
+
+    return reader.of(left(new Error('No 2** response defined, cannot mock')));
   },
 
   negotiateOptionsBySpecificCode(
     httpOperation: IHttpOperation,
     desiredOptions: NegotiationOptions,
     code: string,
-  ): IHttpNegotiationResult {
+  ): Reader<Logger, Either<Error, IHttpNegotiationResult>> {
     // find response by provided status code
-    const responseByForcedStatusCode =
-      findResponseByStatusCode(httpOperation.responses, code) ||
-      createResponseFromDefault(httpOperation.responses, code);
-    if (responseByForcedStatusCode) {
-      try {
-        // try to negotiate
-        return helpers.negotiateOptionsBySpecificResponse(httpOperation, desiredOptions, responseByForcedStatusCode);
-      } catch (error) {
-        // if negotiations fail try a default code
-        try {
-          return helpers.negotiateOptionsForDefaultCode(httpOperation, desiredOptions);
-        } catch (error2) {
-          throw new Error(`${error}. We tried default response, but we got ${error2}`);
-        }
+    return withLogger(logger => {
+      const result = findResponseByStatusCode(httpOperation.responses, code);
+      if (!result) {
+        logger.info(`Unable to find a ${code} response definition`);
+        return createResponseFromDefault(httpOperation.responses, code);
       }
-    }
-    // if no response found under a status code throw an error
-    throw new Error('Requested status code is not defined in the schema.');
+
+      return result;
+    }).chain(responseByForcedStatusCode => {
+      if (responseByForcedStatusCode) {
+        return helpers
+          .negotiateOptionsBySpecificResponse(httpOperation, desiredOptions, responseByForcedStatusCode)
+          .chain(either =>
+            either.fold(
+              () =>
+                helpers
+                  .negotiateOptionsForDefaultCode(httpOperation, desiredOptions)
+                  .map(innerEither =>
+                    innerEither.mapLeft(error => new Error(`${error}. We tried default response, but we got ${error}`)),
+                  ),
+              result => reader.of(right(result)),
```

Comment:
No idea what `reader.of` does. Wanted to read the documentation, here is what it says... https://gcanti.github.io/fp-ts/modules/Reader.ts.html#of-constant

Reply:
Good point, I'm opening a PR right now to make it more clear.

The definition is really simple though

`of: a => new Reader(() => a)`

Is fundamentally a shortcut to create a `Reader` returning a value directly without having to manually write the function.

Think of it as

`jest.fn().mockImplementation(()=> t)` vs `jest.fn().mockReturnValue(t)`

The reason this is required is because composition happens when types align — and so we need to make sure we always return a `Reader` to make the `chain` mechanism work.